pandas/core/internals/managers.py

```diff
@@ -339,8 +338,8 @@ def references_same_values(self, mgr: BaseBlockManager, blkno: int) -> bool:
         Checks if two blocks from two different block managers reference the
         same underlying values.
         """
-        ref = weakref.ref(self.blocks[blkno])
-        return ref in mgr.blocks[blkno].refs.referenced_blocks
+        blk = self.blocks[blkno]
+        return any(blk is ref() for ref in mgr.blocks[blkno].refs.referenced_blocks)
```

Comment:
At the point where `df["col"].index is df.index` is no longer true, we will have to update this, but I assume we can handle that at that point

Reply:
That should work anyway though, since df.col.index will also be in the references since it still shares values with df.index. All shallow copies of the original values should be in there, that’s the elegant think about the tracking mechanism

Reply:
Actually index checks will never get there since it's defined on the manager